// ng-circle-progress-day-countdown/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CircleProgressComponent, CircleProgressOptions } from './circle-progress.component';
export { CircleProgressOptions, CircleProgressComponent } from './circle-progress.component';
var NgCircleProgressModule = (function () {
    function NgCircleProgressModule() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    NgCircleProgressModule.forRoot = function (options) {
        if (options === void 0) { options = {}; }
        return {
            ngModule: NgCircleProgressModule,
            providers: [
                { provide: CircleProgressOptions, useValue: options }
            ]
        };
    };
    return NgCircleProgressModule;
}());
export { NgCircleProgressModule };
NgCircleProgressModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    CircleProgressComponent,
                ],
                exports: [
                    CircleProgressComponent,
                ]
            },] },
];
/**
 * @nocollapse
 */
NgCircleProgressModule.ctorParameters = function () { return []; };
function NgCircleProgressModule_tsickle_Closure_declarations() {
    /** @type {?} */
    NgCircleProgressModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    NgCircleProgressModule.ctorParameters;
}
